File: robot.rb
$placeX = $placeY = -1
$coordinat1 = $coordinat2 = []
$placeD = ""
$points = {
  'A' => [0, 0],
  'B' => [1, 0],
  'C' => [1, 1],
  'D' => [0, 1]
}
$direction = {
  "north" => {"coordinat1" => "D", "coordinat2" => "C", "moveX" => 0, "moveY" => 1, "maxPoint" => 5, "f" => 1 },
  "west"  => {"coordinat1" => "A", "coordinat2" => "D", "moveX" => -1, "moveY" => 0, "maxPoint" => 0, "f" => 0 },
  "south" => {"coordinat1" => "B", "coordinat2" => "A", "moveX" => 0, "moveY" => -1, "maxPoint" => 0, "f" => 1},
  "east"  => {"coordinat1" => "C", "coordinat2" => "B", "moveX" => 1, "moveY" => 0, "maxPoint" => 5, "f" =>  0}
}
class Robot

  def prompt_conversation
    print "> "
  end

  def get_origin(c, c_value)
    until ( !c_value.nil? && Array(0..4).include?(c_value) && (c_value.is_a? Integer))
      prompt_conversation
      puts "Enter "+c+" position"
      c_value = gets.chomp.to_i
    end
    return c_value
  end

  def get_originDirection(d_value)
    puts "Choose direction, NORTH, WEST, SOUTH or EAST"
    d_value = gets.chomp.downcase
    until ( $direction.keys.include?(d_value) )
      if !d_value.nil?
        prompt_conversation
        puts "Invalid input"
      end
      d_value = gets.chomp.downcase
    end
    $placeD = d_value
  end

  def getCoordinats(d)
    # Get the facing direction points
    $coordinat1 = [
      $placeX + ($points[$direction[d]['coordinat1']][0]).to_i,
      $placeY + ($points[$direction[d]['coordinat1']][1]).to_i,
    ]

    $coordinat2 = [
      $placeX + ($points[$direction[d]['coordinat2']][0]).to_i,
      $placeY + ($points[$direction[d]['coordinat2']][1]).to_i,
    ]
  end

  def placeRobot
    puts "Please enter coordinates between 0 to 4."
    $placeX = get_origin("X", nil)
    $placeY = get_origin("Y", nil)
    get_originDirection($placeD)
    getCoordinats($placeD)
  end

  def promptPlace
    puts "Please place the robot first"
    return false
    commands
  end

  def moveRobot
      $moveCoordinat1 = [$coordinat1[0] + $direction[$placeD]['moveX'] , $coordinat1[1] + $direction[$placeD]['moveY']]
      $moveCoordinat2 = [$coordinat2[0] + $direction[$placeD]['moveX'] , $coordinat2[1] + $direction[$placeD]['moveY']]

      if(($moveCoordinat1[$direction[$placeD]['f']] >= 0) && ($moveCoordinat1[$direction[$placeD]['f']] <= 5) )
        $coordinat1 = $moveCoordinat1;
        $coordinat2 = $moveCoordinat2;

        $placeX += $direction[$placeD]['moveX'];
        $placeY += $direction[$placeD]['moveY'];
      else
        puts "Robot cannot move"
      end
  end

  def getIndex
    return ($direction.keys).index($placeD);
  end

  def dKeys
    return $direction.keys
  end

  def leftRobot
    indexD = getIndex
    $placeD = (indexD + 1 == dKeys.size) ? dKeys[0] : dKeys[indexD + 1];
    getCoordinats($placeD);
  end

  def rightRobot
    indexD = getIndex
    $placeD = (indexD-1 == -1) ? dKeys[(dKeys).size - 1] : dKeys[indexD -1];
    getCoordinats($placeD);
  end

  def reportRobot
    puts "#{$placeX}, #{$placeY}, #{$placeD}"
  end

  def isPlaced
    if (Array(0..4).include?($placeX) && Array(0..4).include?($placeY))
      return true
    end
  end

  def commands
    puts "You can control the robot by entering PLACE, MOVE, LEFT, RIGHT and REPORT commands"

    case gets.chomp.upcase
      when "PLACE" then placeRobot
      when "MOVE" then (if isPlaced then moveRobot end)
      when "LEFT" then (if isPlaced then leftRobot end)
      when "RIGHT" then (if isPlaced then rightRobot end)
      when "REPORT" then (if isPlaced then reportRobot end)
    end

    commands
  end

  def init
    prompt_conversation
    puts "Hi! Welcome to the Robot Test."

    puts "First you need to PLACE the robot."

    commands
  end

end
object = Robot.new
object.init

File: assets/js/robot.js
$(document).ready(function(){

  var points = {
    'A' : [0, 0],
    'B' : [1, 0],
    'C' : [1, 1],
    'D' : [0, 1]
  };

  var maxX = 5;
  var maxY = 5;

  var placeX,placeY,placeF,coordinat1,coordinat2, report;

  var direction = {
    "north" : {"coordinat1": "D", "coordinat2": "C", "moveX": 0, "moveY": 1, "maxPoint": 5, "f": 1 },
    "west"  : {"coordinat1": "A", "coordinat2": "D", "moveX": -1, "moveY": 0, "maxPoint": 0, "f": 0 },
    "south" : {"coordinat1": "B", "coordinat2": "A", "moveX": 0, "moveY": -1, "maxPoint": 0, "f": 1},
    "east"  : {"coordinat1": "C", "coordinat2": "B", "moveX": 1, "moveY": 0, "maxPoint": 5, "f":  0}
  };

  var keys = Object.keys(direction);

  $('#place-form').submit(function(e){
    e.preventDefault();
    // Get the form data
    var formArray = $(this).serializeArray();
    placeX = parseInt(formArray[0]['value']);
    placeY = parseInt(formArray[1]['value']);
    placeF = formArray[2]['value'];

    if(placeX > maxX || placeY > maxY) return false;

    getCoordinats(placeF);

    // Make buttons clickable
    $('.directionalButton').each(function(){
      $(this).prop("disabled", false);
    });
    alert("Robot Placed");

  });
  // Move Button
  $('#move').click(function(){
    moveCoordinat1 = [parseInt(coordinat1[0]) + direction[placeF]['moveX'] , parseInt(coordinat1[1]) + direction[placeF]['moveY']];
    moveCoordinat2 = [parseInt(coordinat2[0]) + direction[placeF]['moveX'] , parseInt(coordinat2[1]) + direction[placeF]['moveY']];

    if((moveCoordinat1[direction[placeF]['f']] >= 0) && (moveCoordinat1[direction[placeF]['f']] <= 5) ){
      alert('Robot Moved');
      coordinat1 = moveCoordinat1;
      coordinat2 = moveCoordinat2;

      placeX += direction[placeF]['moveX'];
      placeY += direction[placeF]['moveY'];
    }else{
      alert('Robot can not move');
      return false;
    }
  });

  // Directions of the Robot
  $('#right').click(function(){
    var indexF = keys.indexOf(placeF);
    placeF = (indexF-1 == -1) ? keys[keys.length - 1] : keys[indexF -1];
    getCoordinats(placeF);
  });

  $('#left').click(function(){
    var indexF = keys.indexOf(placeF);
    placeF = (indexF + 1 == keys.length) ? keys[0] : keys[indexF + 1];
    getCoordinats(placeF);
  });

  // Display the Report
  $('#report').click(function(){
    $('#report-board').html(placeX+', '+placeY+', '+placeF);
  });

  function getCoordinats(placeF){
    // Get the facing direction points
    coordinat1 = [
      placeX + parseInt(points[direction[placeF]['coordinat1']][0]),
      placeY + parseInt(points[direction[placeF]['coordinat1']][1])
    ];

    coordinat2 = [
      placeX + parseInt(points[direction[placeF]['coordinat2']][0]),
      placeY + parseInt(points[direction[placeF]['coordinat2']][1])
    ];
  }

});
